Lib/pdb.py

```diff
@@ -624,11 +626,95 @@ def _disable_command_completion(self):
             self.completenames = completenames
         return
 
+    def _exec_in_closure(self, source, globals, locals):
+        """ Run source code in closure so code object created within source
+            can find variables in locals correctly
+
+            returns True if the source is executed, False otherwise
+        """
+
+        # Determine if the source should be executed in closure. Only when the
+        # source compiled to multiple code objects, we should use this feature.
+        # Otherwise, we can just raise an exception and normal exec will be used.
+
+        code = compile(source, "<string>", "exec")
+        if not any(isinstance(const, CodeType) for const in code.co_consts):
+            return False
+
+        # locals could be a proxy which does not support pop
+        # copy it first to avoid modifying the original locals
+        locals_copy = dict(locals)
+
+        locals_copy["__pdb_eval__"] = {
+            "result": None,
+            "write_back": {}
+        }
```

Comment:
Do we need to initialize these, or are there really cases where the `exec` may succeed without these being set?
```suggestion
        locals_copy["__pdb_eval__"] = {}
```

Reply:
In fact, do we need this nested dictionary at all? Or can we just assign to `__pdb_eval__` and `__pdb_write_back__` in the generated code directly, and pop those out of the locals? I think that might make it a bit easier to follow, since it removes a level of nesting.

Reply:
(In fact I think that's what an earlier version of this PR did.)

Reply:
locals will not be updated. locals is only used to initialized the fast variables in the frame and they were never converted back.

Reply:
Without initializing them, we need to do something like `if pdb_eval.pop("result") is None` after. It's not horrible, but I don't really see the benefit here not initializing them.